feat: add Kubernetes pod logs component WIP

mclarke47 opened the pull request:
Fixes https://github.com/backstage/backstage/issues/15753

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [ ] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

hritikch24 commented:
@jamieklassen Please review this pr as logs are needed urgently.

hritikch24 commented:
@mclarke47 i used this branch but it is throwing error while trying to pull logs

mclarke47 commented:
@hritikch24 you might have to send through your cluster config

hritikch24 commented:
> @hritikch24 you might have to send through your cluster config

I am using serviceaccount for authentication which is working for other kubernetes components but not working for logs only.

mclarke47 commented:
@hritikch24  I just merged master into my branch can you try now?

hritikch24 commented:
> @hritikch24 I just merged master into my branch can you try now?

sure.

hritikch24 commented:
> @hritikch24 I just merged master into my branch can you try now?

I tried but i am still facing same issue i.e Client network socket disconnected before secure TLS connection was established"

mclarke47 commented:
@hritikch24 I think let's raise an issue once this is merged, I can't seem to reproduce

hritikch24 commented:
> @hritikch24 I think let's raise an issue once this is merged, I can't seem to reproduce
Ok sure.
I am running it on aws ec2 instance. Is there any host related config which needs to be updated specific for this log module?
Because in network i am getting 500 error.

jamieklassen commented:
I have found at least one bug in the proxy method on KubernetesBackendClient, I'm putting together a PR and then will re-review this against the fixed baseline

hritikch24 reviewed (commented):
Proxy issues

jamieklassen commented:
@hritikch24 #17306

Rugvip commented on plugins/kubernetes/src/api/types.ts:
```diff
@@ -26,6 +26,10 @@ export const kubernetesApiRef = createApiRef<KubernetesApi>({
 });
 
+export const kubernetesProxyApiRef = createApiRef<KubernetesProxyApi>({
+  id: 'plugin.kubernetesproxy.service',
```
```suggestion
  id: 'plugin.kubernetes.proxy-service',
```

Or something like that right? The plugin is still `kubernetes`

Rugvip commented on plugins/kubernetes/src/api/types.ts:
```diff
@@ -49,3 +53,12 @@ export interface KubernetesApi {
+    namespace: string;
+    clusterName: string;
+    containerName: string;
+  }): Promise<string>;
```
```suggestion
  }): Promise<{ text: string }>;
```

Or something like that, makes it a lot easier to evolve this API in the future

Rugvip commented on plugins/kubernetes/api-report.md:
```diff
@@ -399,11 +410,73 @@ const kubernetesPlugin: BackstagePlugin<
 export { kubernetesPlugin };
 export { kubernetesPlugin as plugin };
 
+// Warning: (ae-missing-release-tag) "KubernetesProxyApi" is part of the package's API, but it is missing a release tag (@alpha, @beta, @public, or @internal)
```
Try to keep warnings to a minimum, they do help you catch actual issues

Rugvip commented on .changeset/polite-dingos-train.md:
```diff
@@ -0,0 +1,25 @@
+
+Added Pod logs components for Kubernetes plugin
+
+**BREAKING**: `kubernetesProxyApi` must be defined as an apis in the plugin to use k8s proxy features
```
This isn't true right? There's a factory installed in the plugin so an instance for this will never be missing.

There's plenty of other breaking changes related to components though

Rugvip reviewed (commented):
Nice! 😁 

Couple of nits regarding the API surface, didn't dig too deep into the actual UI changes though

mclarke47 commented:
Can I get a review @Rugvip @jamieklassen ? 🙏

jamieklassen commented:
I have been putting off reviewing this because it involves a lot of parts of the codebase with which I'm not familiar -- all my significant Kubernetes changes have been in the backend. I feel badly because I'm the other project area maintainer besides @mclarke47 so it naturally falls to me to review, and now it's blocked!

I mentioned this to @Rugvip and he said something similar sometimes happens amongst the core maintainers -- the person who creates a PR is often much more knowledgeable about its contents than anybody else. In these cases, they get on a call together. The author walks through the changes and the reviewer has time to ask questions as they go. Let's try this!

jamieklassen commented on .changeset/polite-dingos-train.md:
````diff
@@ -0,0 +1,33 @@
+
+Added Pod logs components for Kubernetes plugin
+
+**BREAKING**: `kubernetesProxyApi` must be defined as an apis in the plugin to use k8s proxy features
+
+```
+...
+export const kubernetesPlugin = createPlugin({
+  id: 'kubernetes',
````
Can we change the verbiage here to mention that this type of setup is required for other plugins (besides the ootb kubernetes plugin) to consume the logs "porcelain"?

mclarke47 replied:
fixed!

jamieklassen commented on plugins/kubernetes/src/api/KubernetesProxyClient.test.ts:
```diff
@@ -0,0 +1,55 @@
+
+    const response = await proxy.getPodLogs(request);
+    await expect(response).toStrictEqual({ text: 'Hello World' });
+    expect(callProxyMock).toHaveBeenCalledTimes(1);
+    expect(callProxyMock.mock.calls[0][0]).toStrictEqual({
+      clusterName: 'some-cluster',
+      init: {
+        method: 'GET',
+      },
+      path: '/api/v1/namespaces/some-namespace/pods/some-pod/log?container=some-container',
+    });
```
You can use the [`toHaveBeenCalledWith` matcher](https://jestjs.io/docs/expect#tohavebeencalledwitharg1-arg2-) here to collapse this into a single cleaner assertion

mclarke47 replied:
fixed!

jamieklassen commented on plugins/kubernetes/src/components/Pods/PodLogs/PodLogs.tsx:
```diff
@@ -0,0 +1,60 @@
+import { usePodLogs } from './usePodLogs';
+
+interface PodLogsProps {
+  previous?: boolean;
```
Let's drop this since it's currently unused; we can explore the use cases later

mclarke47 replied:
fixed!

jamieklassen commented on plugins/kubernetes/src/components/Pods/PodLogs/types.ts:
```diff
@@ -0,0 +1,25 @@
+ * limitations under the License.
+ */
+
+export interface PodLogContext {
```
Maybe we can call this `PodScope` to avoid ambiguous term 'Context' esp in React

mclarke47 replied:
fixed!

jamieklassen commented on plugins/kubernetes/src/components/Pods/PodLogs/types.ts:
```diff
@@ -0,0 +1,25 @@
+  clusterName: string;
+}
+
+export interface ContainerLogContext extends PodLogContext {
```
ditto, maybe `ContainerScope`

jamieklassen reviewed (changes requested):
UX problems we noticed during the review:
* 'YAML' toggle appears on the left-hand side for Pods and on the right-hand side for other types
* when you mouse over the 'X' at the top right of these drawers, the highlight is awkwardly stretched (to tall)

Some non-blocking questions:
* In the Containers section in the KubernetesDrawer, what about `initContainers`? Maybe they don't show up at all?
* Does "Pending" ever signify something other than a pod waiting to start? Maybe a different title for the `PendingPodContent` besides "Pod startup conditions"? "Pod is Pending. Conditions:"

mclarke47 commented:
@jamieklassen  have addressed your comments:

> * 'YAML' toggle appears on the left-hand side for Pods and on the right-hand side for other types

this is fixed

> * when you mouse over the 'X' at the top right of these drawers, the highlight is awkwardly stretched (to tall)

this is fixed

Some non-blocking questions:

> In the Containers section in the KubernetesDrawer, what about initContainers? Maybe they don't show up at all?

These won't show up but I think let's add them in a follow up PR

> Does "Pending" ever signify something other than a pod waiting to start? Maybe a different title for the PendingPodContent besides "Pod startup conditions"? "Pod is Pending. Conditions:"

this is fixed

jamieklassen commented on plugins/kubernetes/src/components/Pods/PodDrawer/ContainerCard.tsx:
```diff
@@ -84,7 +84,7 @@ const ContainerDatetime = ({ prefix, dateTime }: ContainerDatetimeProps) => {
 
 export interface ContainerCardProps {
-  logContext: PodLogContext;
+  logContext: PodScope;
```
Feel free to disagree here, but maybe we rename all the instances of `logContext` to be `scope`?

mclarke47 replied:
fixed

jamieklassen commented on plugins/kubernetes/src/components/Pods/PodLogs/PodLogs.tsx:
```diff
@@ -19,21 +19,18 @@ import { DismissableBanner, LogViewer } from '@backstage/core-components';
 interface PodLogsProps {
-  previous?: boolean;
-  logContext: ContainerLogContext;
+  logContext: ContainerScope;
```
same thing, something like `scope`

mclarke47 replied:
fixed

jamieklassen commented on plugins/kubernetes/src/components/Pods/PodLogs/usePodLogs.ts:
```diff
@@ -16,15 +16,14 @@
 interface PodLogsOptions {
-  logContext: ContainerLogContext;
-  previous?: boolean;
+  logContext: ContainerScope;
```
scope here too

mclarke47 replied:
fixed

jamieklassen reviewed (approved):
I'm OK merging as is, but you might agree with my further naming comments

jamieklassen commented on plugins/kubernetes/src/components/KubernetesDrawer/ManifestYaml.tsx:
```diff
@@ -0,0 +1,57 @@
+        text={jsyaml.dump(object, {
+          replacer: (key: string, value: string): any => {
+            if (!managedFields) {
+              return key === 'managedFields' ? undefined : value;
```
Not sure exactly how `jsyaml` works, but what if I had a key called `managedFields` in my `spec` or somewhere other than `metadata` -- would this thing remove it?

mclarke47 replied:
I put a TODO in for me to fix this
